ibis/backends/tests/tpc/ds/test_queries.py

```diff
@@ -3441,6 +3441,140 @@ def test_74(customer, store_sales, date_dim, web_sales):
     return expr
 
 
+@tpc_test("ds")
+def test_75(
+    catalog_sales,
+    item,
+    date_dim,
+    catalog_returns,
+    store_sales,
+    store_returns,
+    web_sales,
+    web_returns,
+):
+    def _sales(
+        sales,
+        item_sk,
+        sold_date_sk,
+        sales_number,
+        sale_quant,
+        sale_price,
+        returns,
+        rtn_order_number,
+        rtn_item_sk,
+        rtn_quant,
+        rtn_amt,
+    ):
+        return (
+            sales.join(item, item_sk == item.i_item_sk)
+            .join(date_dim, sold_date_sk == date_dim.d_date_sk)
+            .left_join(
+                returns,
+                [
+                    sales_number == rtn_order_number,
+                    item_sk == rtn_item_sk,
+                ],
+            )
+            .filter(_.i_category == "Books")
+            .select(
+                _.d_year,
+                _.i_brand_id,
+                _.i_class_id,
+                _.i_category_id,
+                _.i_manufact_id,
+                sales_cnt=sale_quant - rtn_quant.coalesce(0),
+                sales_amt=sale_price - rtn_amt.coalesce(0.0),
+            )
+        )
+
+    _catalog_sales = _sales(
+        catalog_sales,
+        _.cs_item_sk,
+        _.cs_sold_date_sk,
+        _.cs_order_number,
+        _.cs_quantity,
+        _.cs_ext_sales_price,
+        catalog_returns,
+        catalog_returns.cr_order_number,
+        catalog_returns.cr_item_sk,
+        catalog_returns.cr_return_quantity,
+        catalog_returns.cr_return_amount,
+    )
+
+    _store_sales = _sales(
+        store_sales,
+        _.ss_item_sk,
+        _.ss_sold_date_sk,
+        _.ss_ticket_number,
+        _.ss_quantity,
+        _.ss_ext_sales_price,
+        store_returns,
+        store_returns.sr_ticket_number,
+        store_returns.sr_item_sk,
+        store_returns.sr_return_quantity,
+        store_returns.sr_return_amt,
+    )
+
+    _web_sales = _sales(
+        web_sales,
+        _.ws_item_sk,
+        _.ws_sold_date_sk,
+        _.ws_order_number,
+        _.ws_quantity,
+        _.ws_ext_sales_price,
+        web_returns,
+        web_returns.wr_order_number,
+        web_returns.wr_item_sk,
+        web_returns.wr_return_quantity,
+        web_returns.wr_return_amt,
+    )
+
+    all_sales = (
+        _catalog_sales.union(_store_sales, _web_sales)
```

Comment:
You can't spot the difference because it's incredibly non-obvious and only after about 30 minutes of tinkering did I realize what the problem is: `a UNION b` only yields distinct rows, while `a UNION ALL b` yields every row from `a` and `b`.

In Ibis, `union` defaults to `distinct=False` which produces `UNION ALL`.

Generally speaking `UNION` is *incredibly* rare in practice, partially because it tends to be outrageously expensive to keep a giant hash table of all unique rows hanging around. I'm not 100% sure why this is the default, but I'm guessing it's because it's a rather large performance foot gun.

```suggestion
        _catalog_sales.union(_store_sales, _web_sales, distinct=True)
```

Reply:
Wow.  Thank you.  I stared at that SQL for a loooooong time